src/builtin.cpp

```diff
@@ -3030,6 +3030,69 @@ static int builtin_bg(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
     return res;
 }
 
+/// Helper for builtin_disown
+static int disown_job(parser_t &parser, io_streams_t &streams, job_t *j) {
+    if (j == 0) {
+        streams.err.append_format(_(L"%ls: Unknown job '%ls'\n"), L"bg");
+        builtin_print_help(parser, streams, L"disown", streams.err);
+        return STATUS_BUILTIN_ERROR;
+    }
+
+    // Stopped disowned jobs must be manually signalled; explain how to do so
+    if (job_is_stopped(j)) {
+        streams.err.append_format(
+            _(L"%ls: job %d ('%ls') is stopped. Use 'kill -SIGCONT -%d' to continue it."),
+            L"disown", j->job_id, j->command_wcstr(), j->pgid);
+    }
+
+    if (parser.job_remove(j)) {
+        return STATUS_BUILTIN_OK;
+    } else {
+        return STATUS_BUILTIN_ERROR;
+    }
+}
+
+/// Builtin for removing jobs from the job list
+static int builtin_disown(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
+    int res = STATUS_BUILTIN_OK;
+
+    if (argv[1] == 0) {
+        job_t *j;
+        // Select last constructed job (ie first job in the job queue) that is possible to disown.
+        // Stopped jobs can be disowned (a warning should be printed)
+        // Foreground jobs can be disowned
+        // Even jobs that aren't under job control can be disowned
+        job_iterator_t jobs;
+        while ((j = jobs.next())) {
+            if (j->get_flag(JOB_CONSTRUCTED) && (!job_is_completed(j))) {
+                break;
+            }
+        }
+
+        if (!j) {
+            streams.err.append_format(_(L"%ls: There are no suitable jobs\n"), argv[0]);
+            res = STATUS_BUILTIN_ERROR;
+        } else {
+            res = disown_job(parser, streams, j);
+        }
+    } else {
+        int i;
+        int pid;
+
+        for (i = 1; argv[i]; i++) {
+            pid = fish_wcstoi(argv[i]);
+            if (errno || pid < 0 || !job_get_from_pid(pid)) {
```

Comment:
Now, I'd suggest the following:

- If something that is not an id is specified (i.e. something that isn't a positive integer), that should be a fatal error. No job should be disowned

- If a valid-but-nonexistent id is specified, it should just be skipped. The other jobs should still be disowned.

 The reason for the latter is that the job may have just finished, and the user shouldn't have to rush to get something disowned.